components/kyma-environment-broker/internal/process/deprovisioning/initialisation.go

```diff
@@ -0,0 +1,117 @@
+	CheckStatusTimeout = 3 * time.Hour
+)
+
+type InitialisationStep struct {
+	operationManager  *OperationManager
+	operationStorage  storage.Provisioning
+	instanceStorage   storage.Instances
+	provisionerClient provisioner.Client
+}
+
+func NewInitialisationStep(os storage.Operations, is storage.Instances, pc provisioner.Client) *InitialisationStep {
+	return &InitialisationStep{
+		operationManager:  NewOperationManager(os),
+		operationStorage:  os,
+		instanceStorage:   is,
+		provisionerClient: pc,
+	}
+}
+
+func (s *InitialisationStep) Name() string {
+	return "Deprovision_Init"
+}
+
+func (s *InitialisationStep) Run(operation internal.DeprovisioningOperation, log logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
+	op, err := s.operationStorage.GetProvisioningOperationByInstanceID(operation.InstanceID)
+	if err != nil {
+		log.Errorf("while getting provisioning operation from storage")
+		return operation, time.Second * 10, nil
+	}
+	if op.State == domain.InProgress {
+		log.Info("waiting for provisioning operation to finish")
+		return operation, time.Minute, nil
+	}
+
+	_, err = s.instanceStorage.GetByID(operation.InstanceID)
+	switch {
+	case err == nil:
+		if operation.ProvisionerOperationID == "" {
+			return operation, 0, nil
+		}
+		log.Info("instance being removed, check operation status")
+		return s.checkRuntimeStatus(operation, log)
+	case dberr.IsNotFound(err):
+		return s.operationManager.OperationSucceeded(operation, "instance already deprovisioned")
+	default:
+		log.Errorf("unable to get instance from storage: %s", err)
+		return operation, 1 * time.Second, nil
+	}
+}
+
+func (s *InitialisationStep) checkRuntimeStatus(operation internal.DeprovisioningOperation, log logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
+	if time.Since(operation.UpdatedAt) > CheckStatusTimeout {
+		log.Infof("operation has reached the time limit: updated operation time: %s", operation.UpdatedAt)
+		return s.operationManager.OperationFailed(operation, fmt.Sprintf("operation has reached the time limit: %s", CheckStatusTimeout))
+	}
+
+	instance, err := s.instanceStorage.GetByID(operation.InstanceID)
+	if err != nil {
+		return operation, 10 * time.Second, nil
+	}
+
+	status, err := s.provisionerClient.RuntimeOperationStatus(instance.GlobalAccountID, operation.ProvisionerOperationID)
+	if err != nil {
+		return operation, 1 * time.Minute, nil
+	}
+	log.Infof("call to provisioner returned %s status", status.State.String())
+
+	var msg string
+	if status.Message != nil {
+		msg = *status.Message
+	}
+
+	switch status.State {
+	case gqlschema.OperationStateSucceeded:
+		repeat, err := s.removeInstance(instance.InstanceID)
+		if err != nil || repeat != 0 {
+			return operation, repeat, err
+		}
+		return s.operationManager.OperationSucceeded(operation, msg)
+	case gqlschema.OperationStateInProgress:
+		return operation, 2 * time.Minute, nil
+	case gqlschema.OperationStatePending:
```

Comment:
what is the difference between pending and in progress?
next question: how long does the deprovisioning take? if it is 5 min - retries must not be bigger than half a minute in my opinion. if it takes 1 minute usually, lets retry after 15 sec

Reply:
it takes about 15mins if it's successful so I will set retries to 1min